### PR TITLE
Terra redirect for workspace summary page [risk: low]

### PR DESCRIPTION
**Pull request by davidangb**
* add logic to redirect to Terra, based on the current location in FireCloud
* add config setting to turn redirects on/off - defaulting to off, since redirects would break tests right now
* add a `?fcredir=1` query param to the redirect, to allow for future analytics tracking
* add a Terra redirect for workspace summary, as a proof of the logic

This PR wants - but does not require -  the corresponding `firecloud-develop` PR at broadinstitute/firecloud-develop#1591

work remaining:
* all additional contextual redirects. I'd like to do these in future PRs to keep things small.

-----

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/cljs/main/broadfcui/main.cljs [hunk 1]
```diff
@@ -204,6 +204,15 @@
...
      (let [window-hash (aget js/window "location" "hash")]
+       ;; if Terra redirects are globally enabled via config;
+       ;;   and the current hash has a path handler;
+       ;;   and that path handler has a Terra redirect,
+       ;; then redirect to Terra.
+       (when (config/terra-redirects-enabled)
+         (when-let [handler (nav/find-path-handler window-hash)]
+           (let [{:keys [make-props terra-redirect]} handler]
+             (when terra-redirect
+               (js-invoke (aget js/window "location") "replace" (str (config/terra-url) (terra-redirect (make-props))))))))
```
Comment (breilly2) [inline src/cljs/main/broadfcui/main.cljs]: I like the approach.

Looks like my cljs is rusty. How does `make-props` get its arguments? In this case, it needs `namespace` and `name` in order to construct the URL.

Comment (davidangb) [reply src/cljs/main/broadfcui/main.cljs]: `make-props` gets its arguments based on the regular expression capture matches, as defined by the `regex` in each path definition, e.g. https://github.com/broadinstitute/firecloud-ui/blob/develop/src/cljs/main/broadfcui/page/workspace/details.cljs#L190. The code to calculate the regex matches is at https://github.com/broadinstitute/firecloud-ui/blob/develop/src/cljs/main/broadfcui/nav.cljs#L34.

Comment (breilly2) [reply src/cljs/main/broadfcui/main.cljs]: Ah, thanks. What I was missing is that the `make-props` attribute returned from `find-path-handler` captures the arguments; it's not the same parameterized `make-props` from the call to `nav/defpath`.

File: src/cljs/main/broadfcui/main.cljs [hunk 2]
```diff
@@ -204,6 +204,15 @@
...
      (let [window-hash (aget js/window "location" "hash")]
+       ;; if Terra redirects are globally enabled via config;
+       ;;   and the current hash has a path handler;
+       ;;   and that path handler has a Terra redirect,
+       ;; then redirect to Terra.
+       (when (config/terra-redirects-enabled)
+         (when-let [handler (nav/find-path-handler window-hash)]
+           (let [{:keys [make-props terra-redirect]} handler]
+             (when terra-redirect
+               (js-invoke (aget js/window "location") "replace" (str (config/terra-url) "/?fcredir=1#" (terra-redirect (make-props))))))))
```
Comment (breilly2) [inline src/cljs/main/broadfcui/main.cljs]: Could we also potentially use `fcredir` as a switch to show a "Welcome to FireCloud powered by Terra" message in Terra?

Comment (davidangb) [reply src/cljs/main/broadfcui/main.cljs]: sure! sounds good. That would also be good place in Terra UI to *strip* the `?fcredir=1` from the url so it isn't persistent across "pages".

**Review by breilly2 — CHANGES_REQUESTED** [review]
A couple of things.
1. This is working for in-app navigation but not for direct links, such as bookmarks or opening links in a new tab/window (e.g. http://local.broadinstitute.org/#workspaces/broad-dsde-dev/real_bams)
2. The view starts to render in FC before the redirect happens. Is there any way to interrupt the render flow when we know we're going to redirect to avoid that while the browser is connecting to Terra?

File: src/cljs/main/broadfcui/main.cljs (hunk 2 above)
Comment (breilly2) [reply src/cljs/main/broadfcui/main.cljs]: Agreed!

**Comment by davidangb** [comment]
@breilly2 :
1. direct/bookmark links should  now be working.  The problem was that in `component-will-mount`, we hadn't yet loaded values from `config.json`, so the check to see if redirects are enabled would always fail.  Navigating through the app works because once you've loaded any component in the app, config was loaded, and we could rely on the previously-loaded values.
2. related to the previous bullet ... without a large refactor, we have to load at least one component before redirects will work, so *something* is going to render. How important do you think this is? (the other solution is to eliminate the config check to see if redirects are enabled; I'd  prefer to keep that)

**Comment by breilly2** [comment]
The flicker alone probably isn't a big deal. Combined with the extra API calls that will happen while the view is rendering might be enough to tip in favor of doing this. Those API calls would show up in dashboards as FireCloud usage as well as put unnecessary extra load on the backend services.

I'm thinking a simple redirect component:
```
(react/defc- Redirect
  {:render
   (fn []
     [:div {} "Redirecting..."])})
```
Then extract a `terra-redirect?` boolean and include this somewhere in the `cond` further down:
```
terra-redirect?
[Redirect]
```

**Comment by davidangb** [comment]
Since the redirect code is already in the top-level `App` component, it means refactoring that component entirely  - I'll take a shot and see what kind of effort/risk is involved.

**Review by breilly2 — APPROVED** [review]
👍 Simple and effective. Very nice!

**Pull request merged**